### PR TITLE
Allow referencing previously defined named types

### PR DESCRIPTION
**Pull request by fcoedno**
Once a named type is defined, in order to reuse that same type it is
needed to reference that type by the name (as in the primitive types).

Add support for NamedTypes in both the schema builder and the schema
generator, which is a way of referencing previously defined named types.

As an example, the following snippet wouldn't work, because the schema cannot be parsed:

```php
<?php

declare(strict_types=1);

namespace Acme\Types;

use FlixTech\AvroSerializer\Objects\Schema\Generation\Annotations as SerDe;

/**
 * @SerDe\AvroName("SomeDocument")
 * @SerDe\AvroType("record")
 */
class SomeDocument
{
    /**
     * @SerDe\AvroName("Status")
     * @SerDe\AvroType("enum", attributes={
     *     @SerDe\AvroSymbols({"Undefined", "NotAuthorized", "Canceled"}),
     *     @SerDe\AvroName("Status")
     * })
     */
    private $status;
}
```

```php
<?php

declare(strict_types=1);

namespace Acme\Types;

use FlixTech\AvroSerializer\Objects\Schema\Generation\Annotations as SerDe;

/**
 * @SerDe\AvroType("record")
 * @SerDe\AvroName("SomeDocumentCdc")
 * @SerDe\AvroNamespace("org.acme.types")
 */
class SomeDocumentCdc
{
    /**
     * @SerDe\AvroName("NewDocument")
     * @SerDe\AvroType("record", attributes={
     *     @SerDe\AvroTargetClass("\Acme\Types\SomeDocument")
     * })
     */
    private $newDocument;

    /**
     * @SerDe\AvroName("OldDocument")
     * @SerDe\AvroType("null")
     * @SerDe\AvroType("record", attributes={
     *     @SerDe\AvroTargetClass("\Acme\Types\Some\Document")
     * })
     */
    private $oldDocument;
}
```
With this PR we'll be able to do the following:

```php
<?php

declare(strict_types=1);

namespace Acme\Types;

use FlixTech\AvroSerializer\Objects\Schema\Generation\Annotations as SerDe;

/**
 * @SerDe\AvroType("record")
 * @SerDe\AvroName("SomeDocumentCdc")
 * @SerDe\AvroNamespace("org.acme.types")
 */
class SomeDocumentCdc
{
    /**
     * @SerDe\AvroName("NewDocument")
     * @SerDe\AvroType("record", attributes={
     *     @SerDe\AvroTargetClass("\Acme\Types\SomeDocument")
     * })
     */
    private $newDocument;

    /**
     * @SerDe\AvroName("OldDocument")
     * @SerDe\AvroType("null")
     * @SerDe\AvroType("SomeDocument")
     */
    private $oldDocument;
}
```
```

### CODE REVIEW

**Comment by fcoedno** [comment]
@tPl0ch the CI error seems to be something related to uploading code coverage report, would you have any clues on how to solve that?

File: src/Objects/Schema/Generation/SchemaGenerator.php [hunk 1]
```diff
@@ -123,7 +123,9 @@ private function schemaFromTypes(Type ...$types): Schema
 
                 return $this->applyAttributes($schema, $attributes);
             default:
-                throw new \InvalidArgumentException('$type is not a valid avro type');
+                $schema = Schema::named($type->getTypeName());
```
Comment (tPl0ch) [inline src/Objects/Schema/Generation/SchemaGenerator.php:126]: Do you also have the feeling that this method is doing multiple things - it processes the collection of types and a single type. I believe it makes sense to split these into `schemaFromTypeCollection` and `schemaFromSingleType`.

Additionally I am generally not a huge fan of large `switch`/`case` blocks. One possilbe other way that I have used is using a map of closures instead:

```php
<?php

$typeProcessorMap = [
    TypeName::ENUM => function (SchemaAttributes $attributes) {
        return $this->applyAttributes(Schema::enum(), $attributes);
    },
   // ... more types
];
```

This map can be defined in the constructor of the `SchemaGenerator` class.

File: test/Objects/Schema/Generation/SchemaGeneratorTest.php [hunk 2]
```diff
@@ -45,6 +45,7 @@ public function it_should_generate_an_empty_record()
             ->namespace('org.acme');
 
         $this->assertEquals($expected, $schema);
+        $schema->parse();
```
Comment (tPl0ch) [inline test/Objects/Schema/Generation/SchemaGeneratorTest.php:48]: Why do we need to parse the schema after the expectation? Is there maybe another test assumption/condition missing?

File: test/Objects/Schema/Generation/SchemaGeneratorTest.php [hunk 3]
```diff
@@ -156,9 +159,17 @@ public function it_should_generate_records_containing_records()
                     ->name('SimpleRecord')
                     ->namespace('org.acme')
                     ->field('intType', Schema::int(), Schema\Record\FieldOption::default(42))
+            )
+            ->field(
+                'union',
```
Comment (tPl0ch) [inline test/Objects/Schema/Generation/SchemaGeneratorTest.php:164]: Nit: Should use the name constant.

**Review by tPl0ch — CHANGES_REQUESTED** [review]
Hey @fcoedno,

when seeing your contributions I feel that this library is indeed very valuable to you and that sparks great joy in my heart! ❤️ 

Would you mind looking at some of the comments and questions I have left so that we can get this merged?

Regarding the CI failures, I believe that the failures are either a glitch of GitHub actions themselves, or something is wrong with the action provider. I will take care of this issue.

File: test/Objects/Schema/Generation/SchemaGeneratorTest.php (hunk 2 above)
Comment (fcoedno) [reply test/Objects/Schema/Generation/SchemaGeneratorTest.php:48]: I was not so sure also if it was the best place to do this here. My intention is just to make sure the generated schema object is parseable. There are some checks and validations that will be applied only after we actual do the parsing.

So if anything is wrong, an exception would be raised at this point.

Another idea I had was to create a data provider that would provide the class with the annotations and the expected schema. Then we would have two tests: One to actually see if the generated schema matches the expected one and the other to make sure the generated schema parses without errors.

What do you think?

File: test/Objects/Schema/Generation/SchemaGeneratorTest.php (hunk 3 above)
Comment (fcoedno) [reply test/Objects/Schema/Generation/SchemaGeneratorTest.php:164]: I don't think so. Here the union name does not represent the union type itself, it is just the name of the field. Maybe the best would be to give the field another name to avoid confusion?

Comment (tPl0ch) [reply test/Objects/Schema/Generation/SchemaGeneratorTest.php:164]: The name did indeed confuse me. I like the idea of renaming it to something that indicates an example fieldname rather than the type.

File: test/Objects/Schema/Generation/SchemaGeneratorTest.php (hunk 2 above)
Comment (tPl0ch) [reply test/Objects/Schema/Generation/SchemaGeneratorTest.php:48]: I believe that testing for the failure conditions should be a separate test. That test can use a data provider to provide invalid edge cases and make sure that an exception is thrown. Right now I believe the tests are mixing up different responsibilities - testing that no exception is thrown and the valid parsing of annotations.

**Comment by tPl0ch** [comment]
> @tPl0ch the CI error seems to be something related to uploading code coverage report, would you have any clues on how to solve that?

It seems that this is a [limitation of GitHub Actions](https://github.community/t/make-secrets-available-to-builds-of-forks/16166) that repository secrets are not shared with pipelines running in the forks. I will refactor the actions at a later time and would accept this PR even with this step failing.

File: test/Objects/Schema/Generation/SchemaGeneratorTest.php (hunk 2 above)
Comment (fcoedno) [reply test/Objects/Schema/Generation/SchemaGeneratorTest.php:48]: Idd. Also in this case testing for failures and expect exceptions wouldn't make much sense. Because the parsing and validation of schema is responsibility of the avro-php package. I'm going to remove these "parse" calls

File: src/Objects/Schema/Generation/SchemaGenerator.php (hunk 1 above)
Comment (fcoedno) [reply src/Objects/Schema/Generation/SchemaGenerator.php:126]: Idd. I've refactored it and splitted in two methods. Also, I've extracted the mapping logic into a separate class and also followed your suggestion of creating an associative array with closures.

File: src/Objects/Schema/Generation/TypeMapper.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+declare(strict_types=1);
+
+namespace FlixTech\AvroSerializer\Objects\Schema\Generation;
+
+use FlixTech\AvroSerializer\Objects\Schema;
+use FlixTech\AvroSerializer\Objects\Schema\AttributeName;
+use FlixTech\AvroSerializer\Objects\Schema\TypeName;
+
+class TypeMapper
```
Comment (tPl0ch) [inline src/Objects/Schema/Generation/TypeMapper.php:11]: Nit: This could be a final class

File: src/Objects/Schema/Generation/SchemaGenerator.php [hunk 5]
```diff
@@ -53,78 +59,26 @@ private function generateFromClass(ReflectionClass $class, Type $type): Schema
         return $schema;
     }
 
-    private function schemaFromTypes(Type ...$types): Schema
+    /**
+     * @param array<Type> $types
+     */
+    private function schemaFromTypes(array $types): Schema
```
Comment (tPl0ch) [inline src/Objects/Schema/Generation/SchemaGenerator.php:65]: Could you explain to me why you changed the signature to a simple array over using the variadic type argument?

```php
private function schemaFromTypes(Type ...$types): Schema
```

**Review by tPl0ch — CHANGES_REQUESTED** [review]

File: src/Objects/Schema/Generation/SchemaGenerator.php (hunk 5 above)
Comment (fcoedno) [reply src/Objects/Schema/Generation/SchemaGenerator.php:65]: Simply to avoid unpacking the arrays when passing arguments to this function. But thinking better about this, we loose a more strict type checking by doing so. Let me revert this change.

**Comment by fcoedno** [comment]
Besides fixing the problem of named types itself, I've also found another bug. It was not possible to define extra custom attributes to a record type from the field definition:

```php
<?php

    /**
     * @SerDe\AvroName("simpleField")
     * @SerDe\AvroType("record", attributes={
     *     @SerDe\AvroTargetClass("\FlixTech\AvroSerializer\Test\Objects\Schema\Generation\Fixture\SimpleRecord"),
     *     @SerDe\AvroDoc("This a simple record for testing purposes")
     * })
     */
    private $simpleRecord;
```

**Review by tPl0ch — APPROVED** [review]
LGTM

**Comment by tPl0ch** [comment]
@fcoedno Your work is awesome and much appreciated! ❤️
